File: python-basics/data-structures/queue.py
"""A simple FIFO queue backed by a deque.

This is the Python mirror of the Java QueueUsingArray in
data-structures/queue/. Same FIFO behaviour, idiomatic Python.

Operations and time complexity:
  enqueue(element)  O(1)
  dequeue()         O(1)
  peek()            O(1)
  is_empty()        O(1)
  size()            O(1)

Design note: the Java version is a fixed-capacity circular array that
reuses freed slots and reports when full. A deque is the production
choice here: it grows automatically and pops from the front in O(1),
which a plain list cannot do efficiently.
"""

from collections import deque


class Queue:
    def __init__(self):
        self._items = deque()

    def enqueue(self, element):
        self._items.append(element)

    def dequeue(self):
        if self.is_empty():
            raise IndexError("dequeue from an empty queue")
        return self._items.popleft()

    def peek(self):
        if self.is_empty():
            raise IndexError("peek on an empty queue")
        return self._items[0]

    def is_empty(self):
        return len(self._items) == 0

    def size(self):
        return len(self._items)

    def __str__(self):
        return f"Queue front -> rear: {list(self._items)}"


if __name__ == "__main__":
    queue = Queue()
    queue.enqueue(100)
    queue.enqueue(200)
    queue.enqueue(300)
    print(queue)
    print("Front value:", queue.peek())
    print("Dequeued value:", queue.dequeue())
    print("Size after dequeue:", queue.size())

File: structured-programming/data-structures/queue.cpp
// A simple fixed-capacity circular queue.
//
// This is the C++ mirror of the Java QueueUsingArray in
// data-structures/queue/. Same FIFO behaviour, including the circular
// buffer that reuses freed slots.
//
// Operations and time complexity:
//   enqueue(value)  O(1)
//   dequeue()       O(1)
//   peek()          O(1)
//   isEmpty()       O(1)
//   isFull()        O(1)
//   size()          O(1)

#include <iostream>

using namespace std;

const int MAX_CAPACITY = 100;

class Queue {
private:
    int values[MAX_CAPACITY];
    int capacity;
    int frontIndex;
    int rearIndex;
    int size;

public:
    Queue(int capacity) {
        this->capacity = capacity;
        frontIndex = 0;
        rearIndex = -1;
        size = 0;
    }

    void enqueue(int value) {
        if (isFull()) {
            cout << "The queue is full. I cannot enqueue " << value << "." << endl;
            return;
        }
        rearIndex = (rearIndex + 1) % capacity;
        values[rearIndex] = value;
        size++;
    }

    int dequeue() {
        if (isEmpty()) {
            cout << "The queue is empty. There is no value to dequeue." << endl;
            return -1; // sentinel, mirrors the Java version's null
        }
        int removedValue = values[frontIndex];
        frontIndex = (frontIndex + 1) % capacity;
        size--;
        return removedValue;
    }

    int peek() {
        if (isEmpty()) {
            cout << "The queue is empty. There is no front value." << endl;
            return -1;
        }
        return values[frontIndex];
    }

    bool isEmpty() {
        return size == 0;
    }

    bool isFull() {
        return size == capacity;
    }

    int getSize() {
        return size;
    }

    void printQueue() {
        if (isEmpty()) {
            cout << "Queue: empty" << endl;
            return;
        }
        cout << "Queue front -> rear: ";
        for (int index = 0; index < size; index++) {
            int circularIndex = (frontIndex + index) % capacity;
            cout << values[circularIndex];
            if (index < size - 1) {
                cout << ", ";
            }
        }
        cout << endl;
    }
};

int main() {
    Queue queue(3);
    queue.dequeue();   // empty dequeue
    queue.enqueue(100);
    queue.enqueue(200);
    queue.enqueue(300);
    queue.enqueue(400); // capacity reached: dropped
    queue.printQueue();
    cout << "Front value: " << queue.peek() << endl;
    cout << "Dequeued value: " << queue.dequeue() << endl;
    queue.enqueue(400); // wraps around into the freed slot
    queue.printQueue();
    cout << "Current size: " << queue.getSize() << endl;
    return 0;
}
